input/dpdk/dpdkPortTelemetry.cpp

```diff
@@ -0,0 +1,332 @@
+/**
+ * \file
+ * \brief Implementation of the DpdkPortTelemetry class and related helper functions
+ * \author Pavel Siska <siska@cesnet.cz>
+ * \date 2024
+ */
+/*
+ * Copyright (C) 2024 CESNET
+ *
+ * LICENSE TERMS
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in
+ *    the documentation and/or other materials provided with the
+ *    distribution.
+ * 3. Neither the name of the Company nor the names of its contributors
+ *    may be used to endorse or promote products derived from this
+ *    software without specific prior written permission.
+ */
+
+#include "dpdkPortTelemetry.hpp"
+
+#include <algorithm>
+#include <array>
+#include <iomanip>
+#include <limits>
+#include <numeric>
+#include <sstream>
+#include <stdexcept>
+#include <string>
+#include <vector>
+
+#include <rte_ethdev.h>
+
+namespace ipxp {
+
+static struct rte_eth_dev_info getDeviceInfo(uint16_t portId)
+{
+    struct rte_eth_dev_info devInfo;
+
+    const int ret = rte_eth_dev_info_get(portId, &devInfo);
+    if (ret < 0) {
+        throw std::runtime_error("PortManager::getDeviceInfo() has failed");
+    }
+
+    return devInfo;
+}
+
+static std::string getDeviceNameByPortId(uint16_t portId)
+{
+    std::array<char, RTE_ETH_NAME_MAX_LEN> deviceName;
+
+    const int ret = rte_eth_dev_get_name_by_port(portId, deviceName.data());
+    if (ret < 0) {
+        return "";
+    }
+
+    return {deviceName.data()};
+}
+
+static std::string getRssHashKeyByPortId(uint16_t portId)
+{
+    uint8_t rssHashKeySize = 0;
+    try {
+        rssHashKeySize = getDeviceInfo(portId).hash_key_size;
+    } catch (const std::exception& ex) {
+        return "";
+    }
+
+    std::vector<uint8_t> rssHashKey(rssHashKeySize);
+
+    struct rte_eth_rss_conf rssConf = {};
+    rssConf.rss_key = rssHashKey.data();
+    rssConf.rss_key_len = rssHashKeySize;
+
+    const int ret = rte_eth_dev_rss_hash_conf_get(portId, &rssConf);
+    if (ret < 0) {
+        return "";
+    }
+
+    std::ostringstream oss;
+    for (const auto& byte : rssHashKey) {
+        oss << std::hex << std::setw(2) << std::setfill('0') << static_cast<int>(byte);
+    }
+    return oss.str();
+}
+
+static std::string getRssHashByPortId(uint16_t portId)
+{
+    struct rte_eth_rss_conf rssConf = {};
+    rssConf.rss_key = nullptr;
+    rssConf.rss_key_len = 0;
+
+    const int ret = rte_eth_dev_rss_hash_conf_get(portId, &rssConf);
+    if (ret < 0) {
+        return "";
+    }
+
+    std::vector<std::string> rssHashes;
+
+    if ((rssConf.rss_hf & RTE_ETH_RSS_IPV4) != 0U) {
```

Comment:
I believe that this will only work for DPDK 21.11 and newer. Should we add compatibility for older versions?

Reply:
Added `dpdkCompat.hpp` file with compatible macro definitions